internal/service/db/impl.go

```diff
@@ -0,0 +1,268 @@
+func (*dbService) GetRegistry(
+	_ context.Context,
+) (*toolhivetypes.UpstreamRegistry, string, error) {
+	return nil, "", service.ErrNotImplemented
+}
+
+// ListServers returns all servers in the registry
+func (s *dbService) ListServers(
+	ctx context.Context,
+	opts ...service.Option[service.ListServersOptions],
+) ([]*upstreamv0.ServerJSON, error) {
+	// TODO: implement
+	options := &service.ListServersOptions{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, err
+		}
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(options.Cursor)
+	if err != nil {
+		return nil, err
+	}
+	nextTime, err := time.Parse(time.RFC3339, string(decoded))
+	if err != nil {
+		return nil, err
+	}
+
+	querierFunc := func(querier sqlc.Querier) ([]helper, error) {
+		servers, err := querier.ListServers(
+			ctx,
+			sqlc.ListServersParams{
+				Next: &nextTime,
+				Size: int64(options.Limit),
+			},
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		helpers := make([]helper, len(servers))
+		for i, server := range servers {
+			helpers[i] = listServersRowToHelper(server)
+		}
+
+		return helpers, nil
+	}
+
+	return s.sharedListServers(ctx, querierFunc)
+}
+
+// ListServerVersions implements RegistryService.ListServerVersions
+func (s *dbService) ListServerVersions(
+	ctx context.Context,
+	opts ...service.Option[service.ListServerVersionsOptions],
+) ([]*upstreamv0.ServerJSON, error) {
+	options := &service.ListServerVersionsOptions{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, err
+		}
+	}
+
+	querierFunc := func(querier sqlc.Querier) ([]helper, error) {
+		servers, err := querier.ListServerVersions(
+			ctx,
+			sqlc.ListServerVersionsParams{
+				Name: options.Name,
+				Next: options.Next,
+				Prev: options.Prev,
+				Size: int64(options.Limit),
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		helpers := make([]helper, len(servers))
+		for i, server := range servers {
+			helpers[i] = listServerVersionsRowToHelper(server)
+		}
+
+		return helpers, nil
+	}
+
+	return s.sharedListServers(ctx, querierFunc)
+}
+
+// GetServer returns a specific server by name
+func (s *dbService) GetServerVersion(
+	ctx context.Context,
+	opts ...service.Option[service.GetServerVersionOptions],
+) (*upstreamv0.ServerJSON, error) {
+	options := &service.GetServerVersionOptions{}
+	for _, opt := range opts {
+		if err := opt(options); err != nil {
+			return nil, err
+		}
+	}
+
+	querierFunc := func(querier sqlc.Querier) ([]helper, error) {
+		server, err := querier.GetServerVersion(
+			ctx,
+			sqlc.GetServerVersionParams{
+				Name:    options.Name,
+				Version: options.Version,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		return []helper{getServerVersionRowToHelper(server)}, nil
+	}
+
+	res, err := s.sharedListServers(ctx, querierFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	return res[0], nil
+}
+
+func (s *dbService) sharedListServers(
```

Comment:
nit: could we get some doc on this perhaps in a separate PR? It does quite a bit

Reply:
You're right, I'll document it.